src/createSetupApi.ts

```diff
@@ -0,0 +1,104 @@
+import {
+  BatchInterceptor,
+  HttpRequestEventMap,
+  Interceptor,
+} from '@mswjs/interceptors'
+import { EventMapType, StrictEventEmitter } from 'strict-event-emitter'
+import {
+  DefaultBodyType,
+  RequestHandler,
+  RequestHandlerDefaultInfo,
+} from './handlers/RequestHandler'
+import { LifeCycleEventEmitter } from './sharedOptions'
+import { pipeEvents } from './utils/internal/pipeEvents'
+import { toReadonlyArray } from './utils/internal/toReadonlyArray'
+import { MockedRequest } from './utils/request/MockedRequest'
+
+/**
+ * Generic class for the mock API setup
+ */
+export abstract class SetupApi<TLifecycleEventsMap extends EventMapType> {
+  private readonly initialHandlers: RequestHandler[]
+
+  protected readonly interceptor: BatchInterceptor<
+    Interceptor<HttpRequestEventMap>[],
+    HttpRequestEventMap
+  >
+  protected readonly emitter = new StrictEventEmitter<TLifecycleEventsMap>()
+  protected readonly publicEmitter =
+    new StrictEventEmitter<TLifecycleEventsMap>()
+  protected currentHandlers: RequestHandler[]
+
+  public readonly events: LifeCycleEventEmitter<Record<string | symbol, any>>
+
+  constructor(
+    interceptors: {
+      new (): Interceptor<HttpRequestEventMap>
+    }[],
+    initialHandlers: RequestHandler[],
+  ) {
+    this.interceptor = new BatchInterceptor({
+      name: 'setup-api',
```

Comment:
We would need to somehow parametrize the batch interceptor's `name` so it's different for each child class (`SetupServerApi`, `SetupWorkerApi`). This name affects how the interceptor stores global module overrides, and having multiple batch interceptors with the same name is a no-op.

Reply:
We could pass the name in to the constructor or use a private property that the sub classes override (eg `this.name`) and 
```ts
class SetupServer extends SetupApi {
  private name = "setup-server"
```

If the name doesn't need to be changed except when defining the concrete classes, using the private property might be simplest?

Reply:
I am also thinking of passing the name to the constructor. 
Another option is to just set the `interceptor` in the subclass since the interceptor will not be useful for the `worker` class I think.

Reply:
I like the private property approach but afaik you cannot force that to be set via TypeScript, can you?

Reply:
I was having trouble setting that as private property, so I did the constructor approach. I could make the property abstract to force the subclasses to define it. However, I wouldn't be able to use that property in the constructor, so I would have to create a helper function to initialize the interceptor. If you know a better way, please let me now, I would love to change it!